Function type declarations (#6037)
dart-lang/site-www: src/content/language/functions.md

This example uses an anonymous function.
More about those in the next section.

## Function types

You can specify the type of a function, which is known as a *function type*.
A function type is obtained from a function declaration header by replacing 

> the function name by the word `Function`. Moreover, you are allowed to omit 

Reviewer: ```suggestion
the function name by the keyword `Function`. Moreover, you are allowed to omit 
```

> the names of positional parameters, but the names of named parameters cannot 

Reviewer: ```suggestion
the names of positional parameters, but the names of named parameters can't 
```

be omitted. For example:

<?code-excerpt "misc/lib/language_tour/functions.dart (function-type)"?>
```dart
void greet(String name, {String greeting = 'Hello'}) =>
    print('$greeting $name!');

// Store `greet` in a variable and call it.
void Function(String, {String greeting}) g = greet;
g('Dash', greeting: 'Howdy');
```

:::note
In Dart, functions are first-class objects,
meaning they can be assigned to variables,
passed as arguments, and returned from other functions.

> The [`typedef`][] keyword defines function types explicitly,
> which can be useful for clarity and reusability.

Reviewer: Since `typedef` can be used for other type aliases and the keyword doesn't act alone in defining the type, perhaps let's restructure this to focus on what developers can use it for. How does this sound?

```suggestion

You can use a [`typdef`][] declaration to explicitly name function types,
which can be useful for clarity and reusability.
```

This example uses an anonymous function.
More about those in the next section.

## Function types

We can specify the type of a function, which is known as a *function type*.
It is actually the type of a first-class function object: An expression whose 

> type is a function type can only evaluate to a first-class function object.

Reviewer 2: These two sentences are difficult for me to parse. What do you think about something like "A function type consists of a return type (including `void`) and a list of zero or more parameters."?

Reviewer 3: Lines 272 and 273 describe the semantics of a function type, that is, what it means. Perhaps there's a need to define 'first-class function object', but that's probably difficult to cram into a few words.

A sentence like "A function type consists of ..." describes the syntax of the function type.

So why don't we just have both?

This would fit quite nicely with the next sentence where the syntax is described in a more indirect fashion.

A function type is obtained from a function declaration header by replacing 
the function name by the word `Function`. Moreover, we are allowed to omit 
the names of positional parameters, but the names of named parameters cannot 
be omitted. For example:

<?code-excerpt "misc/lib/language_tour/functions.dart (function-type)"?>
```dart
void printElement(int element) => print(element);

void doPrint(void Function(int) printFunction, {required List<int> list}) {
  list.forEach(printFunction);
}

var list = [1, 2, 3];

// Store `doPrint` in a variable and call it.

> void Function(void Function(int), {required List<int> list}) g = doPrint;

Reviewer 2: Is it possible to use a simpler example here, rather than one which involves declaring the type of a function that receives a function as one of its parameters?

It's a good to mention that you can do this, don't get me wrong, but I'm wondering if there's a simpler example to start with? Maybe something out of the SDK, like Flutter's WidgetBuilder typedef?

Reviewer 3: Perhaps the example could be adjusted to emphasize the simple function type `void Function(int)` which is used in line 284, as well as the more complex one in line 290?

g(printElement, list: list);
```

:::note
In Dart, functions are first-class objects,
meaning they can be assigned to variables,
passed as arguments, and returned from other functions.

> The [`typedef`][] keyword allows defining function types explicitly,

Reviewer 2: nit: "defines" rather than "allows defining"
